gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemBase.java

```diff
@@ -1508,14 +1508,22 @@ private Credential getCredential(Configuration config)
             .withPrefixes(CONFIG_KEY_PREFIXES)
             .get(config, config::get);
 
-    if (isNullOrEmpty(serviceAccountToImpersonate)) {
-      return Optional.empty();
-    }
-
     GoogleCloudStorageOptions options =
         GoogleHadoopFileSystemConfiguration.getGcsFsOptionsBuilder(config)
             .build()
             .getCloudStorageOptions();
+    if (isNullOrEmpty(serviceAccountToImpersonate)) {
+      serviceAccountToImpersonate = getMatchedServiceAccountToImpersonateFromUser(options);
+    }
+
+    if (isNullOrEmpty(serviceAccountToImpersonate)) {
+      serviceAccountToImpersonate = getMatchedServiceAccountToImpersonateFromGroup(options);
+    }
```

Comment:
If you make this methods to return `Optional` then you can chain them:
```java
serviceAccountToImpersonate =
    getMatchedServiceAccountToImpersonateFromUser(options)
        .orElse(getMatchedServiceAccountToImpersonateFromGroup(options))
        .orElse(serviceAccountToImpersonate);
```

Reply:
this cannot be chained like this unfortunately as orElse will return value type not another Optional. But it indeed can be recursively nested.